codegen/module.go

```diff
@@ -550,64 +568,93 @@ func (system *ModuleSystem) ResolveModules(
 
 	resolvedModules := map[string][]*ModuleInstance{}
 
-	for _, className := range system.classOrder {
-		class := system.classes[className]
-		classInstances := []*ModuleInstance{}
+	for _, moduleDirectoryGlob := range system.moduleSearchPaths {
+		moduleDirectoriesAbs, err := filepath.Glob(filepath.Join(baseDirectory, moduleDirectoryGlob))
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error globbing %q", moduleDirectoryGlob)
+		}
+
+		// We don't know until reading the configuration file what the class name will be. For consistency we will
+		// continue to assume that all of the instances that match a glob pattern are the same class name, so
+		// once we've read the first instance we can populate this.
+		className := ""
+
+		for _, moduleDirAbs := range moduleDirectoriesAbs {
+			stat, err := os.Stat(moduleDirAbs)
+			if err != nil {
+				return nil, errors.Wrapf(
+					err,
+					"internal error: cannot stat %q",
+					moduleDirAbs,
+				)
+			}
 
-		for _, dir := range class.Directories {
+			if !stat.IsDir() {
+				// If a *-config.yaml file, or any other metadata file also matched the glob, skip it, since we are
+				// interested only in the containing directories.
+				continue
+			}
 
-			fullInstanceDirectory := filepath.Join(baseDirectory, dir)
-			if class.ClassType == SingleModule {
-				instance, instanceErr := system.readInstance(
-					packageRoot,
+			moduleDir, err := filepath.Rel(baseDirectory, moduleDirAbs)
+			if err != nil {
+				return nil, errors.Wrapf(
+					err,
+					"internal error: cannot make %q relative to %q",
+					moduleDirAbs,
 					baseDirectory,
-					targetGenDir,
-					className,
-					dir,
 				)
-				if instanceErr != nil {
+			}
+
+			instance, instanceErr := system.readInstance(
+				packageRoot,
+				baseDirectory,
+				targetGenDir,
+				moduleDir,
+			)
+			if instanceErr != nil {
+				if _, ok := instanceErr.(inferError); ok {
+					// It could be the case that a glob pattern like "endpoints/*" matches a folder that is later
+					// matched by a longer glob pattern like "endpoints/tchannel/*", so merely print a warning here
+					// rather than failing to generate. It will fail at a later step if no globbing pattern matches
+					// the wanted instance.
+					fmt.Printf("warn: globbing pattern %q: %s\n", moduleDirectoryGlob, instanceErr.Error())
```

Comment:
This warning message seems unnecessary to me, it warns for each dir that does not contain a module config file, which could become quite a clutter when you have many modules nested.